File: api/yd/download.py
from http.server import BaseHTTPRequestHandler
import json
import urllib.parse
import yt_dlp

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        parsed_url = urllib.parse.urlparse(self.path)
        query = urllib.parse.parse_qs(parsed_url.query)
        
        stream_url = query.get('streamUrl', [''])[0]
        video_url = query.get('url', [''])[0]
        fmt = query.get('format', ['mp4'])[0]
        quality = query.get('quality', ['720p'])[0]

        # If we already have a direct stream URL, redirect to it
        if stream_url and stream_url.startswith('http') and 'googlevideo.com' in stream_url:
            self.send_response(302)
            self.send_header('Location', stream_url)
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            return

        # Otherwise, use yt-dlp to extract the direct stream URL and redirect
        if not video_url:
            self.send_response(400)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'url parameter is required'}).encode())
            return

        try:
            # Parse desired height from quality string like "720p"
            desired_height = 720
            try:
                desired_height = int(quality.replace('p', '').replace('kbps', ''))
            except:
                pass

            is_audio = fmt == 'mp3'

            if is_audio:
                ydl_opts = {
                    'quiet': True,
                    'no_warnings': True,
                    'format': 'bestaudio/best',
                }
            else:
                # Try to get the specific quality, fall back to best available
                ydl_opts = {
                    'quiet': True,
                    'no_warnings': True,
                    'format': f'bestvideo[height<={desired_height}]+bestaudio/best[height<={desired_height}]/bestvideo+bestaudio/best',
                }

            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(video_url, download=False)
                
                # Get the direct URL
                direct_url = info.get('url')
                
                if not direct_url:
                    # If merged format, get the video stream URL
                    requested_formats = info.get('requested_formats', [])
                    if requested_formats:
                        if is_audio:
                            # Get audio stream
                            for rf in requested_formats:
                                if rf.get('acodec', 'none') != 'none':
                                    direct_url = rf.get('url')
                                    break
                        else:
                            # Get video stream (first one is usually video)
                            direct_url = requested_formats[0].get('url')
                    
                    if not direct_url:
                        # Last resort: find best matching format from all formats
                        formats = info.get('formats', [])
                        for f in reversed(formats):
                            if f.get('url') and f['url'].startswith('http'):
                                if is_audio and f.get('acodec', 'none') != 'none' and (f.get('vcodec', 'none') == 'none' or not f.get('height')):
                                    direct_url = f['url']
                                    break
                                elif not is_audio and f.get('height') and f.get('vcodec', 'none') != 'none':
                                    if f['height'] <= desired_height:
                                        direct_url = f['url']
                                        break

                if direct_url:
                    self.send_response(302)
                    self.send_header('Location', direct_url)
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.end_headers()
                else:
                    self.send_response(500)
                    self.send_header('Content-Type', 'application/json')
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.end_headers()
                    self.wfile.write(json.dumps({'error': 'Could not extract direct stream URL'}).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

File: api/yd/download.js
import { execFile } from 'child_process';
import { promisify } from 'util';
import path from 'path';
import fs from 'fs';

const execFileAsync = promisify(execFile);

function getYtDlpBinaryPath() {
  const isWin = process.platform === 'win32';
  const localBin = path.resolve(process.cwd(), isWin ? 'bin/yt-dlp.exe' : 'bin/yt-dlp');
  if (fs.existsSync(localBin)) {
    if (!isWin) {
      try { fs.chmodSync(localBin, 0o755); } catch(e) {}
    }
    return localBin;
  }
  return isWin ? 'yt-dlp.exe' : 'yt-dlp';
}

function sanitizeFilename(name) {
  return (name || 'youtube_video')
    .replace(/[^\w\s\-\.\(\)]/gi, '')
    .replace(/\s+/g, '_')
    .substring(0, 100);
}

export default async function handler(req, res) {
  const urlParams = req.query || Object.fromEntries(new URL(req.url || '/', `http://${req.headers.host || 'localhost'}`).searchParams);
  const streamUrl = urlParams.streamUrl;
  const targetUrl = urlParams.url;
  const title = urlParams.title || 'YouTube_Video';
  const format = urlParams.format || 'mp4';
  const isAudioOnly = format === 'mp3' || urlParams.audio === 'audio_only';

  const filename = `${sanitizeFilename(title)}.${isAudioOnly ? 'mp3' : 'mp4'}`;
  let downloadUrl = streamUrl;

  // Resolve targetUrl via bundled yt-dlp binary if streamUrl is missing or incomplete
  if (!downloadUrl && targetUrl) {
    try {
      const binPath = getYtDlpBinaryPath();
      const formatFlag = isAudioOnly ? '-f bestaudio' : '-f bestvideo+bestaudio/best';
      const { stdout } = await execFileAsync(binPath, ['-g', formatFlag, targetUrl], { timeout: 15000 });
      if (stdout) {
        const urls = stdout.trim().split('\n').filter(l => l.startsWith('http'));
        if (urls.length > 0) {
          downloadUrl = urls[0];
        }
      }
    } catch (e) {
      console.warn('yt-dlp stream resolution error in download handler:', e.message);
    }
  }

  if (!downloadUrl) {
    const errPayload = { error: 'streamUrl or valid YouTube url is required' };
    if (res && res.status) return res.status(400).json(errPayload);
    return new Response(JSON.stringify(errPayload), { status: 400, headers: { 'Content-Type': 'application/json' } });
  }

  try {
    const streamRes = await fetch(downloadUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
      }
    });

    if (!streamRes.ok && streamRes.status !== 206) {
      const errPayload = { error: 'Failed to fetch video stream payload from source' };
      if (res && res.status) return res.status(streamRes.status).json(errPayload);
      return new Response(JSON.stringify(errPayload), { status: streamRes.status, headers: { 'Content-Type': 'application/json' } });
    }

    if (res && res.setHeader) {
      res.setHeader('Content-Type', 'application/octet-stream');
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');

      const arrayBuffer = await streamRes.arrayBuffer();
      return res.status(200).send(Buffer.from(arrayBuffer));
    }

    return new Response(streamRes.body, {
      status: 200,
      headers: {
        'Content-Type': 'application/octet-stream',
        'Content-Disposition': `attachment; filename="${filename}"`,
        'Access-Control-Allow-Origin': '*',
        'Cache-Control': 'no-cache, no-store, must-revalidate',
      }
    });

  } catch (err) {
    console.error('Error in download handler:', err);
    const errPayload = { error: 'Download stream failed', details: err.message };
    if (res && res.status) return res.status(500).json(errPayload);
    return new Response(JSON.stringify(errPayload), { status: 500, headers: { 'Content-Type': 'application/json' } });
  }
}
